Mathlib/Tactic/Basic.lean

```diff
@@ -54,3 +54,114 @@ set_option hygiene false in
+      if e.hasLooseBVars then
+        intro1PStep
+        introsDep
+    | _ => ()
+  intro1PStep : TacticM Unit :=
+    liftMetaTactic fun mvarId => do
+      let (_, mvarId) ← Meta.intro1P mvarId
+      pure [mvarId]
+
+example : ∀ a b : Nat, a = b → b = a := by
+  introv h
+  exact h.symm
+
+macro "assumption'" : tactic => `(allGoals assumption)
+
+elab "exacts" "[" hs:term,* "]" : tactic => do
+  for stx in hs.getElems do
+    evalTactic (← `(tactic| exact $stx))
+  evalTactic (← `(tactic| done))
+
+example (n : Nat) : n = n := by
+  induction n
+  exacts [rfl, rfl]
+  exacts []
+
+--TODO : which expr equality to use?
+elab "guardExprEq " r:term " := " p:term : tactic => do
+  let r ← elabTerm r none
+  let p ← elabTerm p none
+  if not (r == p) then throwError "failed: {r} != {p}"
+
+elab "guardTarget" r:term : tactic => do
+  let r ← elabTerm r none
+  let t ← getMainTarget
+  if not (← isDefEq r t) then throwError m!"target of main goal is {t}"
+
+syntax (name := guardHyp) "guardHyp " ident (" : " term)? (" := " term)? : tactic
+@[tactic guardHyp] def evalGuardHyp : Lean.Elab.Tactic.Tactic := fun stx =>
+  match stx with
+  | `(tactic| guardHyp $h $[: $ty]? $[:= $val]?) => do
+    withMainContext do
+      let fvarid ← getFVarId h
+      let lDecl ←
+        match (← getLCtx).find? fvarid with
+        | none => throwError m!"hypothesis {h} not found"
+        | some lDecl => lDecl
+      if let some p ← ty then
+        let e ← elabTerm p none
+        let hty ← instantiateMVars lDecl.type
+        if not (← isDefEq e hty) then throwError m!"hypothesis {h} has type {lDecl.type}"
+      match lDecl.value?, val with
+      | none, some _        => throwError m!"{h} is not a let binding"
+      | some _, none        => throwError m!"{h} is a let binding"
+      | some hval, some val =>
+          let e ← elabTerm val none
+          let hval ← instantiateMVars hval
+          if not (← isDefEq e hval) then throwError m!"hypothesis {h} has value {hval}"
+      | none, none          => ()
+  | _ => throwUnsupportedSyntax
+
+example (n : Nat) : Nat := by
+  guardHyp n : Nat
+  let m : Nat := 1
+  guardHyp m := 1
+  guardHyp m : Nat := 1
+  guardTarget Nat
+  exact 0
+
+elab "matchTarget" t:term : tactic  => do
+  withMainContext do
+    let (val) ← elabTerm t (← inferType (← getMainTarget))
+    if not (← isDefEq val (← getMainTarget)) then
+      throwError "failed"
+
+syntax (name := byContra) "byContra " (colGt ident)? : tactic
+macro_rules
+  | `(tactic| byContra) => `(tactic| (matchTarget Not _; intro))
+  | `(tactic| byContra $e) => `(tactic| (matchTarget Not _; intro $e))
+macro_rules
+  | `(tactic| byContra) => `(tactic| (apply Decidable.byContradiction; intro))
+  | `(tactic| byContra $e) => `(tactic| (apply Decidable.byContradiction; intro $e))
+macro_rules
+  | `(tactic| byContra) => `(tactic| (apply Classical.byContradiction; intro))
+  | `(tactic| byContra $e) => `(tactic| (apply Classical.byContradiction; intro $e))
```

Comment:
```suggestion
  | `(tactic| byContra $[$e]?) => `(tactic| (apply Classical.byContradiction; intro $[$e:ident]?))
```
Not sure if this works, but it might help

Reply:
Not sure why, but it doesn't work :(
I get `expected ')'` at the first `$` of `intro $[$e:ident]?`

Reply:
`intro` uses `*`, not `?`, which are not compatible. Perhaps they should be.